### ISSUE TITLE
Turning Generac off, or Hat that can control relays

### ISSUE THREAD

**Issue by RalphBSz** [jgyates/genmon#909]
tl;dr: Is there a pi hat that has 4 extra relay outputs, for turning the Generac off and on during a power outage? Or is there a better solution?

We have a Generac with the Evolution 1.0 controller. And I want to be able to do remote monitoring and control of it. We live in an area (20 minutes from Silicon Valley!) where power outages are frequent and prolonged (this year we had at least 38 days of power outages already), so economizing propane and generator runtime is necessary.

For monitoring the generator, things seem easy: I already have several other Raspberry Pi's in production monitoring other things around the house. The GenMon software seems perfect, and I already tried it once on the generator with a USB-to-RS232 adapter and a temporary power cord, and it runs great. To do a real install, I'll need some more permanent hardware, namely a hat with a RS232 output and 12V to 5V converter. Fortunately, @skipfire = pintsize.me makes some which will fit the bill. There is one slight issue here: When we lose power, we also lose internet, even if our generator is running: the upstream phone company switching infrastructure for the DSL line doesn't have good backup batteries. So I have a 5V powered cell phone modem (made by SixFab) installed on it, and I'll need to set up some scripts to switch the internet connection from WiFi to the cell modem. The only problem is whether the 5V regulator on skipfire's hat has enough current for the cell modem; that just needs some testing. So far, so good. By the way, I like the big LED on the hat, as it allows me to have a "user interface" so a human can see (from a window without getting wet) what's happening.

But now comes the real question. As we all "know", there is no way to turn the generator off and back on when the power is out. Which is the same as saying that to my knowledge nobody has ever discovered any serial interface commands that allow doing that. There is some logical justification for this: Generac doesn't like the generator to be stopped when under load without a cooldown period, nor does it like to be started under load, and that makes a little bit of sense. Honestly, this really is a minor issue: Our normal load is less than 1kW, and on a 16kW generator that load barely causes the engine and generator to warm up. But this is Generac's choice. On the other hand, I will turn the generator off at night and during the middle of the day, but I'm too lazy to walk out to it (typically in storms with heavy rain), turn the breaker off and then press the "off" button.

So here's my idea. If one opens the Evolution controller, one can see the wiring for the three off/on/auto buttons on the keypad. It would be easy to add a tiny PC board there, with 3 little reed relays to spoof the buttons. The relay for the "off" button has to be NC (normally closed), since disconnecting the button turns the generator off, the other two are NO. And to address the concern that the generator needs to cool down and then restart without load, I can add a big 70A or 100A contactor to the AC output (between breaker and transfer switch), wired with a 12V NC relay. Like that, an improved version of GenMon could turn the generator off by first disconnecting the 240V load for a minute, then spoofing the off button.

So for this, I would need a hat that has 4 drivers for small relays. Given that they are galvanically isolated, there is no need for an optocoupler, just a transistor array to drive them from the RPi's GPIO pins. And they can be really small reed relays with low current consumption, in particular if driven by the 12V line.

Here are my two questions for discussion: Is my idea for controlling the generator with 4 relays insane? And anyone know of a hat that has RS-232 converter, 5V supply from 12V, LED driver, and 4 outputs for small 12V relays?

**Comment by jgyates** [comment]
Do not know how this would apply to the PintSize.me Hat, but in ***some*** instances you can run multiple Hats on a pi, as long as the GPIO does not conflict. This thread has some discussion on the topic of multiple hats:

https://forums.raspberrypi.com/viewtopic.php?t=141645

Regarding genmon and this functionality, the Evoluiont Liquid cooled models allow Manual/Off/Auto buttons to be set via modbus, however these commands do not work on Evolution Air Cooled models. 

Adding functionality to control relays is easy on the software side so if there were documentation that others could follow on how to accomplish this I could easily add this feature in the software or possibly incorporate an Add On program that would allow this to occur.

**Comment by skipfire** [comment]
Warning - every cell HAT I have seen will be incompatible with Pis running GenMon regardless of the HAT.  Pi only exposes a single serial port which GenMon needs, but the cell modem also needs those pins.  I'd love to hear about a cell HAT that uses alternate pins, but have yet to find one.  There are some USB options though.

The PintSize.Me HAT has a couple of features that could help.  The onboard 12v to 5v buck circuit is design capable of up to 3A output to the Pi, the HAT itself consumes a tiny amount of power so the remainder is available or Pi, Pi's USB, and on full size HATs the fan 5v pin.  On the full size HAT, there is a 5-pin header to the left of the fan slot which has ground and 3.3v along with 3 GPIO pins that line up with the default for one of the GenMon addons that allows you to control start/stop/transfer with external buttons or other GPIO type controls.  That said, it is possible as you said to stack HATs as long as there is no or meaningless overlap with GPIO pins between HATs.  For HATs 1.8 and higher the pin use is as followed, but some may not be in use at some versions/form factors (HAT vs pHAT).
7 - 1-Wire (temperature sensors (HATs only; has cutoff switch on board)
8  - Serial TX (every version of the HAT ever made)
10 - Serial RX (every version of the HAT ever made)
11 - 5-pin header Stop (HATs only)
13 - 5-pin header Run (HATs only)
15 - 5-pin header Xfer (HATs only)
16 - tri color Green (has cutoff switch on board)
18 - tri color Red (has cutoff switch on board)
19 - Button (has cutoff switch on board)
21 - Fan ground transistor control
23 - tri color Blue (NanoPi, has cutoff switch on board)
26 - tri-color Blue (HATs/pHATs, has cutoff switch on board)

8 & 9 are unavoidably locked, without a fan, you could probably use 21 by another HAT.  11, 13, 15 could be used by another HAT as long as you don't use those header pins on this HAT; 16, 18, 19, 26 can all be cutoff with onboard switches, 7 is the default 1-Wire port and could be used for 1-Wire across multiple boards (this does have the appropriate pull resistor onboard though so be aware and the entire 1-Wire circuit (including the pull resistor) can be switched off on the board.

Lastly, PIntSize.Me does do custom HATs as well, use the contact form to contact us directly.  Some of the changes from v1.7 to on the v2.x HATs were inspired from a previous custom version of the GenMon HAT that was requested.

**Comment by jgyates** [comment]
If you want to use a cellular connection, @skipfire is correct that cell HAT boards do use the Pi serial port so you would need to use either USB serial [like this Add On](https://github.com/jgyates/genmon/wiki/1----Software-Overview#gensmspy-optional) does or use [serial over TCP](https://github.com/jgyates/genmon/wiki/Appendix-F----Serial-over-IP), which would not use a local serial port but require a serial over TCP device to be installed on the same LAN segment as the pi (and have the LAN up and powered during an outage).

The existing genmon add on that uses cell presently only uses the cell network for SMS. It does not support a network connection over cellular or inbound messaging.

**Comment by RalphBSz** [comment]
Thank you for the fast replies! Multiple hats is a nice solution; I could breadboard a hat that has just the 4 relay drivers, and stack it underneath @skipfire's hat. Or use the full-size hat and use the 3 pins that are already on connectors: In reality, there is no need to ever remote-control the "on" button, "off" and "auto" is all that's really needed. That would be a nice testing ground, and from there one might proceed to a custom hat.

It's nice to know that the voltage converter can handle 3A; that will be more than sufficient to run a Pi and the cell modem.

The SixFab cellular modems can run (and actually prefer to run) with just a single USB connection, rather than using the Pi's serial port on the GPIO. On that USB they can expose a virtual serial port (for AT modem commands), and also a virtual ethernet port for full internet connectivity. While they have a hat form factor, they really don't need to be stacked on top of a pi, as they can get power from the USB line. They're found at www.sixfab.com, and happen to be a local company here in Silicon Valley.

I'm going to use the cell modem likely as a full internet connection, not just to use the existing GenMon SMS component: If power is out AND home internet is out, the generator-control pi would temporarily act as a "hub" and collect other information from around the house and export it to the cloud for me; that has all been prototyped, and getting it to production quality is currently mostly waiting for weatherproof wiring and all the hardware components.

Again, thank you for the fast answers!

**Comment by skipfire** [comment]
@RalphBSz is your Evolution an air cooled or liquid cooled?  I'm not sure relays will help you for what you have described, though I hope you find a way to do so.  During an outage you cannot turn the generator off with software, and on the air cooled units at least the switch is a physical 3 position rocker with manual-off (the controller doesn't override that), manual-on (again, the controller does not override that) and auto (the controller picks when to start and stop), but no known command will stop the generator in an outage.  If propane management is your goal, you may have to find a way to trick the generator into thinking there is power when there isn't.

The 1-Wire, 5-pin header, and fan are all current HAT only options and are not currently on the pHATs or legacy HAT.

If you do figure it out, having stuff pulling power from the starter battery without either utility power or generator power may drain the battery too much to start, so you may also want to consider some sort of solar battery charger that could be added to the mix if you will want the generator off for significant periods of time.  If you are running it for an hour every few hours of an outage to keep the fridge cold that may suffice but I'm not sure.

Whenever you get your end solution I do hope you post it back here, you are not the first wanting to have remote propane management but I haven't heard of anyone getting it working.

**Comment by RalphBSz** [comment]
It's air-cooled, the 16 kW model, bought in 2014. It has the Evolution controller version 1 (which on the front just says "Evolution", not "Evolution 2.0"; that's the one that already has three pushbuttons at the bottom for off/auto/on (not a big rocker switch), but unlike the version 2 is not WiFi-prepped. The three pushbuttons are easy to find by taking the controller apart: on and auto are momentary contact NO, and off is a momentary contact NC. That makes perfect sense: If the controller is damaged (like cable harness torn), the generator turns off, which is fail safe.

So my plan is to go into the controller, and put three tiny reed relays in parallel / series with those three push buttons. We've had the controller apart already, and I know what wires to connect to, and found a place for a tiny PC board.

On the battery issue: A RPi Zero W on idle uses about 100-150 mA (measured) at 5V, which works out to 45-65 mA at 12V. The starter battery of a Generic is a group 26, which has about 35 Ah capacity (they are really spec'ed as reserve capacity), to it has enough capacity for about 500 hours = 4 weeks of running the RPi until it is fully discharged. Obviously, one doesn't want to fully discharge it (since then the generator won't start), but given that in normal use the generator runs at least twice a day, this is just not an issue. Also, I think on the Evolution controller GenMon can monitor the battery voltage, so the human can run the generator long before one gets into the danger zone. If I had to build my own hat, I would probably add a tiny A/D converter to it, and monitor voltage and current on the RPi, but that's just having belt and suspenders. I'm really hoping to not have to do my own hat, since doing it (netlist, PCB layout, fabbing, repeat a few times) will cost me many weekends.

Once I have it up and running, I'll definitely report here. And if all this works good, you'll hear suggestions for functionality to add to the next generation hat.

**Comment by skipfire** [comment]
@RalphBSz I'd recommend a Zero2 over a Zero W as there is a huge difference in processing power and that may be needed with the cell modem, though a 3A+ would likely be a better option as it has slightly better performance, more case and cooling option, more HATs (since it's fullsize), USB port already on there without an OTG adapter, fairly low power requirements, and 5GHz WiFi support.  

As for the extra functions you need, take a look at https://sequentmicrosystems.com/products/four-relays-four-inputs-for-raspberry-pi.  It's got relays and it says it can read voltage as well, but I've never used it so I can't speak to it more than the description on their site.

**Comment by RalphBSz** [comment]
I'm hoping that the current Zero W can do it, because it has OS, software and scripts already set up. Plus, wait times to buy a Zero2 are long. I already had GenMon running on the current one with no issues.

That board is overkill, since the relays are all going to go into the generator: three tiny reed relays into the Evolution controller, one 12V industrial control relay to drive the 70A contactor that's going to be bolted on next to the breaker. All I need is four transistor drivers with flyback diodes for the relays. I'm thinking whether I should add an opto-coupler; probably will. That's the kind of board I can solder up in one evening using a standard Pi breadbord.

**Comment by jgyates** [comment]
I am going to close this thread, but feel free to post any follow up question or status here.
